pkg/controllers/state/statenode.go

```diff
@@ -150,6 +158,54 @@ func (in *StateNode) Pods(ctx context.Context, c client.Client) ([]*v1.Pod, erro
 	return nodeutils.GetPods(ctx, c, in.Node)
 }
 
+// IsDisruptable returns an error if the StateNode cannot be disrupted
+// This checks all associated StateNode internals, node labels, and do-not-disrupt annotations
+// on both the pod and the node.
+// IsDisruptable takes in a recorder to emit events on the nodeclaims when the state node is not a candidate
+//
+//nolint:gocyclo
+func (in *StateNode) IsDisruptable(ctx context.Context, kubeClient client.Client) error {
+	if in.Node == nil || in.NodeClaim == nil {
+		return fmt.Errorf("state node doesn't contain both a node and a nodeclaim")
+	}
+	if !in.Initialized() {
+		return fmt.Errorf("state node isn't initialized")
+	}
+	if in.MarkedForDeletion() {
+		return fmt.Errorf("state node is marked for deletion")
+	}
+	// skip the node if it is nominated by a recent provisioning pass to be the target of a pending pod.
+	if in.Nominated() {
+		return fmt.Errorf("state node is nominated for a pending pod")
+	}
+	if _, ok := in.Annotations()[v1beta1.DoNotDisruptAnnotationKey]; ok {
+		return fmt.Errorf("disruption is blocked through the %q annotation", v1beta1.DoNotDisruptAnnotationKey)
+	}
+	// check whether the node has all the labels we need
+	for _, label := range []string{
+		v1beta1.CapacityTypeLabelKey,
+		v1.LabelTopologyZone,
+		v1.LabelInstanceTypeStable,
+		v1beta1.NodePoolLabelKey,
+	} {
+		if _, ok := in.Labels()[label]; !ok {
+			return fmt.Errorf("state node doesn't have required label %q", label)
+		}
+	}
+	pods, err := in.Pods(ctx, kubeClient)
```

Comment:
Something to keep in mind: If we use this call here and don't store the pods: The pods that we are going to pull to build the candidate are going to be different from the pods that we are using to evaluate whether the StateNode is disruptable. This may not cause a problem, but we should think through race conditions here

Reply:
I guess we'd need to return the list of pods in ValidateDisruptable if we think that this is a problem. Realistically, I'm not sure it would change that much, considering that it would only modify the disruption cost.

Reply:
Mainly thinking about differences between candidacy and the pods that we actually use for scheduling; though, I think you're right that it probably makes literally no difference, because having to re-grab the pods just means that we will be more correct with our data, it doesn't make any difference from the perspective of whether we choose to work with the node or not